drivers/t5838/t5838_aad_mode.c

```diff
@@ -0,0 +1,486 @@
+	struct t5838_aad_drv_data *drv_data = dev->data;
+	int err;
+
+	/** Make sure there is no PDM transfer in progress, and we can take clock signal */
+	struct t5838_drv_data *pdm_data = drv_cfg->pdm_dev->data;
+	if (pdm_data->active) {
+		LOG_ERR("Cannot write to device while pdm is active");
+		return -EBUSY;
+	}
+	if (drv_data->aad_unlocked == false) {
+		err = prv_aad_unlock_sequence(dev);
+		if (err < 0) {
+			LOG_ERR("error writing aad unlock sequence, err: %d", err);
+			return err;
+		}
+	}
+
+	err = prv_multi_reg_write(dev, write_array, write_array_size);
+	if (err < 0) {
+		LOG_ERR("prv_t5838_multi_reg_write, err: %d", err);
+		return err;
+	}
+
+	err = prv_enter_sleep_with_AAD(dev);
+	if (err < 0) {
+		LOG_ERR("prv_t5838_reg_write, err: %d", err);
+		return err;
+	}
+
+	pdm_data->aad_child_dev = dev;
+	/**Configure interrupts */
+	if (drv_data->cb_configured == false) {
+		drv_data->int_handled = false;
+		gpio_init_callback(&drv_data->wake_cb, prv_wake_cb_handler, BIT(drv_cfg->wake.pin));
+
+		err = gpio_add_callback(drv_cfg->wake.port, &drv_data->wake_cb);
+		if (err) {
+			LOG_ERR("gpio_add_callback, err: %d", err);
+			return err;
+		}
+		drv_data->cb_configured = true;
+		g_drv_data = drv_data;
+		g_drv_cfg = drv_cfg;
+	}
+	err = gpio_pin_interrupt_configure_dt(&drv_cfg->wake, GPIO_INT_EDGE_TO_ACTIVE);
+	if (err) {
+		LOG_ERR("gpio_pin_interrupt_configure_dt, err: %d", err);
+		return err;
+	}
+	return 0;
+}
+
+int t5838_aad_a_mode_set(const struct device *dev, struct t5838_aad_a_conf *aadconf)
+{
+	struct t5838_aad_drv_data *drv_data = dev->data;
+	int err;
+
+	struct t5838_address_data_pair write_data[] = {
+		{T5838_REG_AAD_MODE, 0x00},
+		{T5838_REG_AAD_A_LPF, aadconf->aad_a_lpf},
+		{T5838_REG_AAD_A_THR, aadconf->aad_a_thr},
+		{T5838_REG_AAD_MODE, T5838_AAD_SELECT_A},
+	};
+	err = prv_aad_mode_set(dev, write_data, ARRAY_SIZE(write_data));
+	if (err < 0) {
+		LOG_ERR("prv_aad_mode_set, err: %d", err);
+		return err;
+	}
+	drv_data->aad_enabled_mode = T5838_AAD_SELECT_A;
+	return 0;
+}
+
+int t5838_aad_d1_mode_set(const struct device *dev, struct t5838_aad_d_conf *aadconf)
+{
+	struct t5838_aad_drv_data *drv_data = dev->data;
+	int err;
+	struct t5838_address_data_pair write_data[] = {
+		{T5838_REG_AAD_MODE, 0x00},
+		{T5838_REG_AAD_D_FLOOR_HI, (aadconf->aad_d_floor >> 8) & 0x1F},
+		{T5838_REG_AAD_D_FLOOR_LO, aadconf->aad_d_floor & 0xFF},
+		{0x2C, 0x32},
+		{T5838_REG_AAD_D_REL_PULSE_MIN_LO, aadconf->aad_d_rel_pulse_min & 0xFF},
+		{T5838_REG_AAD_D_ABS_REL_PULSE_MIN_SHARED,
+		 ((aadconf->aad_d_abs_pulse_min >> 4) / 0xF0) |
+			 ((aadconf->aad_d_rel_pulse_min >> 8) & 0x0F)},
+		{T5838_REG_AAD_D_ABS_PULSE_MIN_LO, aadconf->aad_d_abs_pulse_min & 0xFF},
+		{T5838_REG_AAD_D_ABS_THR_LO, aadconf->aad_d_abs_thr & 0xFF},
+		{T5838_REG_AAD_D_ABS_THR_HI, (aadconf->aad_d_abs_thr >> 8) & 0x1F},
+		{T5838_REG_AAD_D_REL_THR, aadconf->aad_d_rel_thr},
+		{T5838_REG_AAD_MODE, T5838_AAD_SELECT_D1},
+	};
```

Comment:
Is this really necessary? I think we don't really gain much since all these 3 functions do now is build this sequence and pass it into another function that does configuration. I think it might just add unnecessary layer that won't really help with either readability or performance.